engine/execution/ingestion/engine.go

```diff
@@ -1220,6 +1240,37 @@ func (e *Engine) logExecutableBlock(eb *entity.ExecutableBlock) {
 	}
 }
 
+func (e *Engine) initWorkQueueLimitChannel() {
+	if e.workQueueLimitCh == nil {
+		e.workQueueLimitSize = DefaultMaxWorkQueueSize
+		e.workQueueLimitCh = make(chan struct{}, e.workQueueLimitSize)
+	}
+}
+
+func (e *Engine) deinitWorkQueueLimitChannel() {
+	if e.workQueueLimitCh != nil {
+		close(e.workQueueLimitCh)
+		e.workQueueLimitCh = nil
+	}
+}
+
+func (e *Engine) blockingMarkOneBlockAddedToWorkQueue() {
+	if e.workQueueLimitCh != nil {
+		// NOTE: since this could be blocking when chan is full, this function should
+		//		 not be called inside mempool.Run() which is globally locked itself.
+		e.workQueueLimitCh <- struct{}{}
+		e.metrics.ExecutionBlockAddedToWorkQueue()
+	}
+}
```

Comment:
is it possible that this causes a deadlock, all the blocks in the queue not be executable since they are all missing a parent block?

Reply:
Good catch @ramtinms. Logically yes it is possible. When that happens, what could be the largest number of blocks being stuck in the waiting to be executed status? The limit size here is supposed to be very large though, but if a missing parent block can exist in arbitrary height, we will have to address this concern.

Also when we reload blocks, could that happen that certain parent block is missing? If not, we can be conservative that this limit will only be applied at block reloading